README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# roughsf <img src="man/figures/logo.png" align="right" width="120px"/>

<!-- badges: start -->

[![R-CMD-check](https://github.com/schochastics/roughsf/workflows/R-CMD-check/badge.svg)](https://github.com/schochastics/roughsf/actions)
[![CRAN
status](https://www.r-pkg.org/badges/version/roughsf)](https://CRAN.R-project.org/package=roughsf)
<!-- badges: end -->

Using the java script library [rough.js](https://roughjs.com/) to draw
sketchy, hand-drawn-like maps  
(*Checkout [ggrough](https://xvrdm.github.io/ggrough/) for turning
general ggplot objects into sketchy drawings and
[roughnet](https://github.com/schochastics/roughnet) for networks*)

![](man/figures/example.png)

## Installation

You can install the developers version of roughsf with:

``` r
# install.packages("remotes")
remotes::install_github("schochastics/roughsf")
```

## Usage

`roughsf()` takes a list of sf objects as input. The layers are drawn in
order as they appear in the list. MULTI\* geometries are not supported
and should be casted beforehand. Aesthetics need to be specified
explicitly as columns in the sf objects.

The following attributes are supported for POLYGONS:

-   *fill* fill color
-   *color* stroke color
-   *stroke* stroke size
-   *fillstyle* one of “hachure”, “solid”, “zigzag”, “cross-hatch”,
    “dots”, “dashed”, “zigzag-line”
-   *fillweight* thickness of fillstyle (between 0 and 1)
-   *hachureangle* angle of hachure lines
-   *hachuregap* gap between two hachure lines

The following attributes are supported for LINESTRINGS:

-   *color* stroke color
-   *stroke* stroke size

The following attributes are supported for POINTS:

-   *color* color of point
-   *size* size of point
-   *label* label to be added (optional)
-   *label_pos* position of label relative to point: (c)enter, (n)orth,
    (e)ast, (s)outh, (w)est (optional)

Default values are used if one of the attributes is not found.

**Values need to be specified explicitly and cannot be inferred from
other columns** (This needs to be done in a preparatory step)

``` r
library(sf)
ger <- rnaturalearth::ne_countries(scale = "medium", country = "Germany", returnclass = "sf")
ger <- st_cast(ger, "POLYGON")
ger$fill <- "#CD2626"
ger$stroke <- 2
ger$fillweight <- 0.5

# MULTIPOLYGON (and also MULTILINESTRING) are not supported
ger <- st_cast(ger, "POLYGON")

cities <- data.frame(name = c("Berlin", "Munich", "Hamburg", "Cologne"))
cities$geometry <- st_sfc(
  st_point(c(13.4, 52.5200)), st_point(c(11.582, 48.1351)),
  st_point(c(9.9937, 53.5511)), st_point(c(6.9603, 50.9375))
)
cities <- st_sf(cities)
st_crs(cities) <- "+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0"
cities$size <- 15
cities$color <- "#000000"
cities$label <- cities$name
cities$label_pos <- "e"

roughsf::roughsf(list(ger, cities),
                 title = "Sketchy Map of Germany", caption = "drawn by @schochastics",
                 title_font = "48px Pristina", font = "30px Pristina", caption_font = "30px Pristina",
                 roughness = 1, bowing = 1, simplification = 1,
                 width = 800, height = 1000, 
)
```

![](man/figures/germany.png)

Plots can be saved as png with the `save_roughsf()` function.

# More Examples

![](man/figures/africa.png)

![](man/figures/pop_dens_ger.png)


NEWS.md
# roughsf 1.0.0

* Added a `NEWS.md` file to track changes to the package.


inst/htmlwidgets/roughsf.js
function drawPoly(rc, s) {
  rc.path(s.xy, {
     roughness: s.roughness,
     bowing: s.bowing,
     simplification: s.simplification,
     fill: s.fill,
     fillStyle: s.fillstyle,
     hachureGap: 4,
     fillWeight: s.fillweight,
     stroke: s.color,
     strokeWidth: s.size
   });
}

function drawLine(rc, s) {
  rc.path(s.xy, {
     roughness: s.roughness,
     bowing: s.bowing,
     stroke: s.color,
     strokeWidth: s.size
   });
}

function drawText(rc,ctx,s) {
  if(s.pos==="c"){
    ctx.textAlign = "center";
    ctx.textBaseline = "middle";
    ctx.fillText(s.label,Number(s.x), Number(s.y));
  }
  if(s.pos==="n"){
    ctx.textAlign = "center";
    ctx.textBaseline = "bottom";
    ctx.fillText(s.label,Number(s.x), Number(s.y)-0.5 * Number(s.size));
  }
  if(s.pos==="s"){
    ctx.textAlign = "center";
    ctx.textBaseline = "top";
    ctx.fillText(s.label,Number(s.x), Number(s.y)+0.5 * Number(s.size));
  }
  if(s.pos==="e"){
    ctx.textAlign = "left";
    ctx.textBaseline = "middle";
    ctx.fillText(s.label,Number(s.x)+0.5 * Number(s.size), Number(s.y));
  }
  if(s.pos==="w"){
    ctx.textAlign = "right";
    ctx.textBaseline = "middle";
    ctx.fillText(s.label,Number(s.x)-0.5 * Number(s.size), Number(s.y));
  }
}

function drawPoint(rc,ctx, s) {
   rc.circle(Number(s.x), Number(s.y), Number(s.size), {
     roughness: s.roughness,
     bowing: s.bowing,
     fill: s.fill,
     fillStyle: "solid",
     hachureGap: 4,
     fillWeight: 1,
     stroke: s.color,
     strokeWidth: 1
   });

   if(s.label!==""){
    drawText(rc,ctx,s)
   }

}

HTMLWidgets.widget({

  name: 'roughsf',

  type: 'output',

  factory: function(el, width, height) {

    return {
      renderValue: function(x) {

        // Create Canvas element in DOM
        var canvas = document.createElement("canvas");
        canvas.setAttribute("id", x.id);
        canvas.setAttribute("width", width);
        canvas.setAttribute("height", height);
        el.appendChild(canvas);

        // Insert rough canvas in the new canvas element
        const rc = rough.canvas(document.getElementById(x.id));

        // Create context for text shape
        const c = document.getElementById(x.id);
        var ctx = c.getContext("2d");
        ctx.font = x.font;

        //draw boarder
        //rc.rectangle (0, 0, width, height,{strokeWidth:2,roughness:3});
        x.data.map(function(s) {
          if(s.shape === "POLYGON"){
            drawPoly(rc, s);
          }
          if(s.shape === "LINESTRING"){
            drawLine(rc, s);
          }

          if(s.shape === "POINT"){
            drawPoint(rc,ctx,s);
          }

          if(s.shape === "TEXT"){
            drawText(rc,ctx,s);
          }

          if(s.shape === "TITLE"){
            ctx.font = x.title_font;
            drawText(rc,ctx,s);
            ctx.font = x.font;
          }

          if(s.shape === "CAPTION"){
            ctx.font = x.caption_font;
            drawText(rc,ctx,s);
            ctx.font = x.font;
          }

        });
      },
      resize: function(width, height) {
        // TODO
      }
    };
  }
});


R/roughsf.R
#' Create a rough map
#' @description plot a sf map using rough.js
#' @param layers an sf object or a list of sf object. each object should only contain one type of geometry.
#' @param roughness numeric vector for roughness of lines
#' @param bowing numeric vector for bowing of lines
#' @param simplification simplify drawings (remove points from objects)
#' @param font font size and font family for labels
#' @param title optional title of the map
#' @param title_font font size and font family for title
#' @param caption optional caption of the map
#' @param caption_font font size and font family for caption
#' @param width width
#' @param height height
#' @param elementId DOM id
#' @param chunk_name markdown specific
#' @details
#' The following attributes are supported for POLYGONS:
#' * \emph{fill} fill color
#' * \emph{color} stroke color
#' * \emph{stroke} stroke size
#' * \emph{fillstyle} one of "hachure", "solid", "zigzag", "cross-hatch", "dots", "dashed", "zigzag-line"
#' * \emph{fillweight} thickness of fillstyle (between 0 and 1)
#' * \emph{hachureangle} angle of hachure lines
#' * \emph{hachuregap} gap between two hachure lines
#'
#' The following attributes are supported for LINESTRINGS:
#' * \emph{color} stroke color
#' * \emph{stroke} stroke size
#'
#' The following attributes are supported for POINTS:
#' * \emph{color} color of point
#' * \emph{size} size of point
#' * \emph{label} label to be added (optional)
#' * \emph{label_pos} position of label relative to point: (c)enter, (n)orth, (e)ast, (s)outh, (w)est (optional)
#'
#' Default values are used if one of the attributes is not found.
#'
#' The result of a roughsf call can be printed to file with `save_roughsf()`
#' @references
#' More details on roughjs can be found on https://github.com/rough-stuff/rough/wiki
#' @export
roughsf <- function(layers,
                    roughness = 1, bowing = 1, simplification = 1,
                    font = "30px Arial",
                    title = NULL, title_font = "30px Arial",
                    caption = NULL, caption_font="30px Arial",
                    width = NULL, height = NULL, elementId = NULL,chunk_name = "canvas") {

  if("sf"%in%class(layers)){
    layers <- list(layers)
  }
  n_layers <- length(layers)

  layer_types <- lapply(layers,sf::st_geometry_type)
  if(any(sapply(layer_types,function(l) length(unique(l)))>1)){
    stop("each layer must only contain one geometry type.")
  }
  layer_types_char <- sapply(layer_types,function(l) as.character(l[1]))
  if(any(layer_types_char%in%c("MULTIPOLYGON","MULTILINESTRING"))){
    stop("MULTIPOLYGONS and MULTILINESTRINGS are not supported. Use `sf::st_cast()` first.")
  }

  coords_list <- lapply(layers,function(l) sf::st_coordinates(l))
  n_pts <- c(0,cumsum(sapply(coords_list,nrow)))
  coords <- do.call("rbind",lapply(coords_list,function(xy)xy[,1:2]))
  if(is.null(width)){
    width <- 800
    coords[,1] <- normalise(coords[,1], to = c(100,700))
  } else{
    coords[,1] <- normalise(coords[,1], to = c(width*0.1,width*0.9))
  }
  if(is.null(height)){
    height <- 600
    coords[,2] <- normalise(coords[,2], to = c(500,100))
  } else{
    coords[,2] <- normalise(coords[,2], to = c(height*0.9,height*0.1))
  }

  for(i in 1:(length(n_pts)-1)){
    coords_list[[i]][,1:2] <- coords[(n_pts[i]+1):n_pts[i+1],]
  }

  rough_lst <- vector("list",n_layers)
  for(i in 1:n_layers){
    if(layer_types_char[i]=="POLYGON"){
      rough_lst[[i]] <- prepare_polygon(layers[[i]],coords_list[[i]])
    } else if(layer_types_char[i]=="LINESTRING"){
      rough_lst[[i]] <- prepare_linestring(layers[[i]],coords_list[[i]])
    } else if(layer_types_char[i]=="POINT"){
      rough_lst[[i]] <- prepare_points(layers[[i]],coords_list[[i]])
    }
  }

  rough_df <- do.call("rbind",rough_lst)

  if(!is.null(title)){
    title_df <- data.frame(xy="",x=width/2,y=50,shape="TITLE",color="black",
                           fill="",fillstyle="", hachureangle = NA, hachuregap = NA,
                           size=NA,fillweight="",label=title,pos="c")
    rough_df <- rbind(rough_df,title_df)
  }

  if(!is.null(caption)){
    caption_df <- data.frame(xy="",x=width/2,y=height*.95,shape="CAPTION",color="black",
                           fill="",fillstyle="", hachureangle = NA, hachuregap = NA,
                           size=NA,fillweight="",label=caption,pos="c")
    rough_df <- rbind(rough_df,caption_df)
  }
  rough_df$roughness <- roughness
  rough_df$bowing <- bowing
  rough_df$simplification <- simplification

  x <- list(
    data=jsonlite::toJSON(rough_df),
    font=font,
    title_font=title_font,
    caption_font=caption_font,
    id=chunk_name
  )

  # create widget
  htmlwidgets::createWidget(
    name = 'roughsf',
    x = x,
    width = width,
    height = height,
    package = 'roughsf',
    elementId = elementId
  )
}

normalise <- function (x, from = range(x), to = c(0, 1))
{
  x <- (x - from[1])/(from[2] - from[1])
  if (!identical(to, c(0, 1))) {
    x <- x * (to[2] - to[1]) + to[1]
  }
  x
}

prepare_polygon <- function(object,coords){
  if(!"fill" %in% names(object)){
    object[["fill"]] <- "black"
  }

  if(!"color" %in% names(object)){
    object[["color"]] <- "black"
  }

  if(!"stroke" %in% names(object)){
    object[["stroke"]] <- 1
  }

  if(!"fillstyle" %in% names(object)){
    object[["fillstyle"]] <- "hachure"
  }

  if(!"fillweight" %in% names(object)){
    object[["fillweight"]] <- 0.5
  }

  if(!"hachureangle" %in% names(object)){
    object[["hachureangle"]] <- 41
  }

  if(!"hachuregap" %in% names(object)){
    object[["hachuregap"]] <- 4*object[["stroke"]]
  }

  nobj <- nrow(object)#max(coords[,4])
  mobj <- 4
  path_string <- rep("",nobj)
  coords_obj <- split(coords[,1:2],coords[,mobj])

  for(i in 1:nobj){
    xy <- matrix(coords_obj[[i]],ncol=2)
    path_string[i] <- paste0("M ",paste0(apply(xy,1,paste0,collapse=" "),collapse=" L "))
    path_string[i] <- paste0(path_string[i]," z")
  }
  data.frame(
    xy  = path_string,
    x = NA,
    y = NA,
    shape = "POLYGON",
    color = object[["color"]],
    fill  = object[["fill"]],
    fillstyle = object[["fillstyle"]],
    size = object[["stroke"]],
    fillweight = object[["fillweight"]],
    hachureangle = object[["hachureangle"]],
    hachuregap = object[["hachuregap"]],
    label="",
    pos="")
}

prepare_linestring <- function(object,coords){
  if(!"color" %in% names(object)){
    object[["color"]] <- "black"
  }

  if(!"stroke" %in% names(object)){
    object[["stroke"]] <- 1
  }

  nobj <- nrow(object)#max(coords[,4])
  mobj <- 3
  path_string <- rep("",nobj)
  coords_obj <- split(coords[,1:2],coords[,mobj])

  for(i in 1:nobj){
    xy <- matrix(coords_obj[[i]],ncol=2)
    path_string[i] <- paste0("M ",paste0(apply(xy,1,paste0,collapse=" "),collapse=" L "))
  }
  data.frame(
    xy  = path_string,
    x = NA,
    y = NA,
    shape = "LINESTRING",
    color = object[["color"]],
    fill  = "",
    fillstyle = "",
    size = object[["stroke"]],
    fillweight = "",
    hachureangle = NA,
    hachuregap = NA,
    label="",
    pos="")
}

prepare_points <- function(object,coords){
  if(!"color" %in% names(object)){
    object[["color"]] <- "black"
  }

  if(!"size" %in% names(object)){
    object[["size"]] <- 15
  }

  if(!"label" %in% names(object)){
    object[["label"]] <- ""
  }

  if(!"label_pos" %in% names(object)){
    object[["label_pos"]] <- "c"
  }

  data.frame(
    xy  = "",
    x = coords[,1],
    y = coords[,2],
    shape = "POINT",
    color = "",
    fill  = object[["color"]],
    fillstyle = "solid",
    size = object[["size"]],
    fillweight = "",
    hachureangle = NA,
    hachuregap = NA,
    label=object[["label"]],
    pos=object[["label_pos"]])
}

prepare_label <- function(object){

}

#<a href="https://www.freepik.com/photos/background">Background photo created by aopsan - www.freepik.com</a>
